Check trip existence

binh-dam-ibigroup opened the pull request:
### Checklist

- [X] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [X] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [na] The description lists all applicable issues this PR seeks to resolve
- [na] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR causes the `MonitoredTripController` to verify that a same-day, non-realtime itinerary exists for all monitored days for a new or existing `MonitoredTrip`. The verification is done using a 7-day window that starts from the date of the saved OTP query. Same-day means from 3am local-time to 3am the following day for the OTP request date. If there is no same-day itinerary, a 400 error is returned. Upon check, `MonitoredTripController` replaces the itinerary provided in the request body with a non-realtime equivalent.

A new `ItineraryUtils` and `ItineraryExistenceChecker` classes (and accompanying tests) break down the core logic for the itinerary checks.

### Things to finalize:

- name for the `itinerarycheck` endpoint,
- whether the `itinerarycheck` endpoint should return all days at once or should be broken out to check individual days (I have not noticed lags in the combined response, so it has not been a primary concern for now).

binh-dam-ibigroup committed:
feat(TripExistenceChecker): Add class to check trip existence.

binh-dam-ibigroup committed:
feat(MonitoredTripController): Prevent saving trip if an itinerary does not exist on all monitored d

binh-dam-ibigroup committed:
refactor(TripExistenceChecker): Prevent null arg in ctor.

binh-dam-ibigroup committed:
refactor: Various refactors.

binh-dam-ibigroup committed:
improvement(MonitoredTripController): Replace trip itinerary with eqv verified from OTP.

binh-dam-ibigroup committed:
improvement(ItineraryUtils): Consider 12am-2:59am to be previous day.

binh-dam-ibigroup committed:
refactor(ItineraryUtils): Remove getResponseForDate() (use map instead)

binh-dam-ibigroup committed:
docs(ItineraryUtils): Tweak comments.

binh-dam-ibigroup committed:
improvement(ItineraryUtils): Compensate a day for OTP plans departing 12-3am.

binh-dam-ibigroup committed:
docs: Tweak comments.

binh-dam-ibigroup committed:
refactor(ItineraryExistenceChecker): Make class return (Otp)Response. Cache response in OtpDispatche

binh-dam-ibigroup committed:
refactor(OtpDispatcherResponse): Prevent serializing Response (fix CI build error)

evansiroky commented on src/main/java/org/opentripplanner/middleware/models/MonitoredTrip.java:
```diff
@@ -279,5 +283,23 @@ public boolean delete() {
         return Persistence.monitoredTrips.removeById(this.id);
     }
+
+    /**
+     * @return the zone id from the trip 'from' location.
+     */
+    public ZoneId tripZoneId() {
+        Optional<ZoneId> fromZoneId = getZoneIdForCoordinates(from.lat, from.lon);
+        if (fromZoneId.isEmpty()) {
+            String message = String.format(
+                "Could not find coordinate's (lat=%.6f, lon=%.6f) timezone for monitored trip %s",
+                from.lat,
+                from.lon,
+                id
+            );
+            throw new RuntimeException(message);
+        } else {
+            return fromZoneId.get();
+        }
+    }
```
Please replace this with the `getTimezoneForTargetLocation` method that I made in #75 (see [here](https://github.com/ibi-group/otp-middleware/pull/75/files#diff-a3ee245c312c932d083bd39feecf69f3R331)) so that we get the timezone from the proper location depending on whether the trip is arrive by or depart at.

binh-dam-ibigroup replied:
Fixed in https://github.com/ibi-group/otp-middleware/pull/77/commits/a2e1ec3403a6fa2f002dcfd1c59f89594a6ca3a8

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java:
```diff
@@ -56,4 +107,30 @@ private void verifyBelowMaxNumTrips(String userId, Request request) {
+        ItineraryExistenceChecker itineraryChecker = new ItineraryExistenceChecker(OtpDispatcher::sendOtpPlanRequest);
+        try {
+            ItineraryExistenceChecker.Result checkResult = itineraryChecker.checkAll(ItineraryUtils.getItineraryExistenceQueries(trip));
+            if (!checkResult.allItinerariesExist) {
+                logMessageAndHalt(
+                    request,
+                    HttpStatus.BAD_REQUEST_400,
+                    "An itinerary does not exist for some of the monitored days for the requested trip."
+                );
+            }
```
In order for the UI to be able to show which days of the week the trip is possible, shouldn't this return some kind of map of the days of the week that the trip is possible?

binh-dam-ibigroup replied:
This method is called on PUT or POST. But the next step is indeed to create an endpoint for the UI to report the checks... Should that be in this PR or a separate one?

binh-dam-ibigroup replied:
I guess I'll go ahead and build the endpoint for the UI too.

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java:
```diff
@@ -0,0 +1,160 @@
+            1597388399000L // August 14, 2020 02:59:59 AM, considered to be Aug 13.
+        );
+        testItineraryDepartsSameDay("1:23", true,
+            1597215600000L, // August 12, 2020 3:00:00 AM
+            1597291199000L, // August 12, 2020 11:59:59 PM
```
I'd actually expect both of these to return false. The logic is that the request date is still August 13th, so one would not expect an itinerary with a start time on August 12th to show up at all. Despite August 12th being the same service day at the time, it is not the same request date.

evansiroky replied:
Actually, the second one might make sense for a trip that was "arriving by" 1:23am.

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java:
```diff
@@ -0,0 +1,185 @@
+    /**
+     * @return true if the itinerary's startTime is one the same day as the day of the specified date and time.
+     */
+    public static boolean itineraryDepartsSameDay(Itinerary itinerary, String date, String time, ZoneId zoneId) {
```
All of this method is oriented around an assumption that the request date and time is relative to a "depart at" kind of query. There needs to be logic related to an "arrive by" query too. What about a hypothetical long trip that was arriving by 3:01am that had to depart at 11:59pm the previous day?

binh-dam-ibigroup replied:
I added the arrive-by tests in https://github.com/ibi-group/otp-middleware/pull/77/commits/0aba7da8c43c1e2b26da517ee70d6f821b9a933b.

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    @Test
+    public void testItineraryDepartsSameDay() {
```
I like these tests, but think we need some more for trips that "arrive by".

binh-dam-ibigroup replied:
I added the arrive-by tests in https://github.com/ibi-group/otp-middleware/pull/77/commits/0aba7da8c43c1e2b26da517ee70d6f821b9a933b

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java:
```diff
@@ -0,0 +1,185 @@
+     * The ui_activeItinerary query parameter is used to determine which itinerary to use,
+     * TODO/FIXME: need a trip resemblance check to supplement the ui_activeItinerary param used in this function.
+     */
+    public static void updateTripWithVerifiedItinerary(MonitoredTrip trip, List<Itinerary> verifiedItineraries) throws URISyntaxException {
```
This method can result in a completely different itinerary being saved than what the user selected. We should discuss things a little more so that we're on the same page of what to do when a user wants to save an itinerary that is only made possible due to unique realtime data. For example, say a bus A is running early and is able to get to a bus stop just before Bus B leaves because Bus B is late. That result won't ever show up in these additional checks here. We probably need some kind of way to determine if an itinerary is typically possible at all or was made possible by the current day's realtime conditions.

evansiroky reviewed (changes requested):
See comments. I think we need to have a work session around this topic and others.

binh-dam-ibigroup committed:
refactor(MonitoredTrip): Use timezoneForTargetLocation() from #75.

binh-dam-ibigroup committed:
refactor(ItineraryUtils): Refactor for checking itinerary same-day arrival.

br648 commented on src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java:
```diff
@@ -27,11 +35,54 @@ public MonitoredTripController(String apiPrefix) {
     MonitoredTrip preCreateHook(MonitoredTrip monitoredTrip, Request req) {
         verifyBelowMaxNumTrips(monitoredTrip.userId, req);
         monitoredTrip.initializeFromItinerary();
+        ItineraryExistenceChecker.Result checkResult = checkItineraryExistence(monitoredTrip, req);
```
Is it worth creating E2E tests for this, similar to ApiUserFlowTest?

binh-dam-ibigroup replied:
Cast your votes for adding e2e tests.

br648 reviewed (commented):
Looking good so far. I understand that additional work is still to be done under this PR, so will review again once that is complete.

binh-dam-ibigroup committed:
feat(ItineraryCheckController): Add endpoint that checks itinerary existence.

The endpoint checks all days of the week at once.

binh-dam-ibigroup committed:
refactor(ItineraryUtilsTest): Refactor tests.

binh-dam-ibigroup committed:
refactor: Refactor src classes.

binh-dam-ibigroup committed:
Merge branch 'dev' into check-trip-existence

binh-dam-ibigroup committed:
Merge branch 'trip-monitor-timing' into check-trip-existence

binh-dam-ibigroup committed:
refactor(Itinerary): Make test pass.

binh-dam-ibigroup committed:
revert: Revert 0805cd2.

binh-dam-ibigroup committed:
refactor(OtpDispatcherResponse): Remove response caching (caused tests to fail)

binh-dam-ibigroup committed:
Merge branch 'check-trip-existence' of https://github.com/ibi-group/otp-middleware into check-trip-existence

binh-dam-ibigroup committed:
refactor(ItineraryUtils): Use parseQueryParams from MonitoredTrip.

binh-dam-ibigroup committed:
refactor: Refactor various files.

binh-dam-ibigroup committed:
fix(MonitoredTripController): Return trip existence message when adding invalid day to existing trip

binh-dam-ibigroup commented:
I'm going ahead and submit this for review. Note that this PR is now dependent on merging #75 as it has some utility code used in this PR. To make the review easier, I'm setting the base branch to that of #75. I'm also adding the BLOCKED label pending merging of #75.

br648 commented on src/main/java/org/opentripplanner/middleware/controllers/api/ItineraryCheckController.java:
```diff
@@ -0,0 +1,121 @@
+                if (r.getValue().plan != null) {
+                    // Only keep same-day itineraries.
+                    List<Itinerary> sameDayItineraries = ItineraryUtils.getSameDayItineraries(r.getValue().plan.itineraries, trip, dateString);
+                    if (sameDayItineraries.size() != 0) {
```
Could replace with !sameDayItineraries.isEmpty()

br648 commented on src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java:
```diff
@@ -36,6 +44,12 @@ MonitoredTrip preCreateHook(MonitoredTrip monitoredTrip, Request req) {
+        // Replace the provided trip's itinerary with a verified, non-real-time version of it.
+        if (checkResult != null) {
+            updateTripWithVerifiedItinerary(monitoredTrip, req, checkResult.labeledResponses);
+        }
```
According to Intellij, the preCreateHook and preUpdateHook contain duplicate code. I think this should be refactored into a single method?

binh-dam-ibigroup replied:
Good catch! Updated in https://github.com/ibi-group/otp-middleware/pull/77/commits/d331144c173deeb3e2e5cc8542f174d42312ff80 along with your other suggestions.

br648 commented on src/main/java/org/opentripplanner/middleware/otp/OtpDispatcherResponse.java:
```diff
@@ -90,4 +99,34 @@ public OtpDispatcherResponse clone() {
 //        clonedObject.responseBody = this.responseBody;
 //        return clonedObject;
     }
+
```
Not strictly part of this PR, but what is happening with the clone method above this? I think the comments should at least be removed.

binh-dam-ibigroup replied:
I don't know... Those changes are from commit 38dafa4b71983ae5f2b8f47bc6e7568007ba5fb1 by @landonreed.

br648 commented on src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java:
```diff
@@ -0,0 +1,190 @@
+     */
+    public static boolean isSameDay(Itinerary itinerary, String date, String time, ZoneId zoneId, boolean checkArrival) {
+        // TODO: Make SERVICEDAY_START_HOUR an optional config parameter.
+        final int SERVICEDAY_START_HOUR = 3;
```
I think this should be SERVICE_DAY_START_HOUR

br648 commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryExistenceCheckerTest.java:
```diff
@@ -0,0 +1,79 @@
+    private static HashMap<String, OtpDispatcherResponse> queryToResponse;
+
+    /** Mock (significantly abbreviated) OTP response when an itinerary is not found. */
+    private final static String MOCK_RESPONSE_WITH_ERROR = "{\"requestParameters\":{},\"error\":{\"id\":404,\"msg\":\"No trip found...\",\"message\":\"PATH_NOT_FOUND\",\"noPath\":true},\"debugOutput\":{}}";
```
Perhaps change to JSON ObjectNodes?

evansiroky replied:
Or import a file from the resources folder.

binh-dam-ibigroup replied:
Good idea! I updated the class to use the error response file that is already in the resources folder. See https://github.com/ibi-group/otp-middleware/pull/77/commits/d331144c173deeb3e2e5cc8542f174d42312ff80.

br648 reviewed (approved):
This looks good to me. Some minor items, but approving.

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java:
```diff
@@ -0,0 +1,190 @@
+    public static Map<String, String> getQueriesFromDates(Map<String, String> params, List<String> dates) {
+        // Create a copy of the original params in which we change the date.
+        Map<String, String> paramsCopy = new HashMap<>(params);
+        Map<String, String> result = new HashMap<>();
```
Just my personal preference, `result` could be named `queryParamsByDate`.

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryExistenceCheckerTest.java:
```diff
@@ -0,0 +1,79 @@
+            TEST_RESOURCE_PATH + "persistence/planResponse.json"
+        );
+        queryToResponse = new HashMap<>();
+        URI uri = URI.create("http://www,example.com");
```
typo of a comma in the URL, but I guess it doesn't affect anything else.

binh-dam-ibigroup committed:
refactor: Address PR comments.

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/ItineraryCheckController.java:
```diff
@@ -0,0 +1,121 @@
+            trip.initializeFromItineraryAndQueryParams();
+
+            ItineraryExistenceChecker itineraryChecker = new ItineraryExistenceChecker(OtpDispatcher::sendOtpPlanRequest);
+            ItineraryExistenceChecker.Result checkResult = itineraryChecker.checkAll(ItineraryUtils.getItineraryExistenceQueries(trip, true), trip.isArriveBy());
```
Line is too long, please break up.

evansiroky commented on src/main/java/org/opentripplanner/middleware/otp/OtpDispatcherResponse.java:
```diff
@@ -90,4 +99,30 @@ public OtpDispatcherResponse clone() {
+                Optional<ZoneId> fromZoneId = getZoneIdForCoordinates(plan.from.lat, plan.from.lon);
+                if (fromZoneId.isPresent()) {
+                    for (Itinerary itinerary : plan.itineraries) {
+                        if (ItineraryUtils.isSameDay(itinerary, requestDate, requestTime, fromZoneId.get(), checkArrival)) {
```
Line is too long, please break up.

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java:
```diff
@@ -0,0 +1,190 @@
+        final int SERVICE_DAY_START_HOUR = 3;
+        DateTimeFormatter dateFormatter = DateTimeFormatter.ofPattern(DEFAULT_DATE_FORMAT_PATTERN);
+
+        ZonedDateTime startDate = ZonedDateTime.ofInstant(itinerary.getStartOrEndTime(checkArrival).toInstant(), zoneId);
```
The `Itinerary#getStartOrEndTime` is only ever called here. I'd recommend just extracting the logic and inline it here.

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java:
```diff
@@ -0,0 +1,190 @@
+     * @param checkArrival true to check the itinerary endtime, false to check the startTime.
+     * @return true if the itinerary's startTime or endTime is one the same day as the day of the specified date and time.
+     */
+    public static boolean isSameDay(Itinerary itinerary, String date, String time, ZoneId zoneId, boolean checkArrival) {
```
The `checkArrival` variable is used here, but in the `Itinerary#getStartOrEndTime` it's then called `isArrival` and in `MonitoredTrip` there is a method called `isArriveBy`. In OTP there is simply an `arriveBy` request parameter. Let's be consistent and try to use `arriveBy` everywhere.

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/ItineraryExistenceChecker.java:
```diff
@@ -0,0 +1,66 @@
+     *                       label that will be used in Result.labeledResponses for easy identification.
+     * @return An object with a map of results and summary of itinerary existence.
+     */
+    public Result checkAll(Map<String, String> labeledQueries, boolean checkArrival) {
```
I'd recommend renaming `chechArrival` to `tripIsArriveBy`.

binh-dam-ibigroup replied:
Good suggestion.

evansiroky commented on src/main/java/org/opentripplanner/middleware/otp/response/Itinerary.java:
```diff
@@ -98,6 +97,18 @@ public void clearAlerts() {
+        return isArrival ? endTime : startTime;
+    }
+
+    public void setStartOrEndTime(Date startOrEnd, boolean isArrival) {
```
This is only ever used during testing. Please move this logic to a test class.

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/ItineraryExistenceChecker.java:
```diff
@@ -0,0 +1,66 @@
+            this.labeledResponses = labeledResponses;
+        }
+    }
+}
```
This file is small enough that it'd probably be fine to merge it into the `ItineraryUtils` class.

binh-dam-ibigroup replied:
Merged in https://github.com/ibi-group/otp-middleware/pull/77/commits/91812109313baeba91b81c2fea66fb1fa954cca9.

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/ItineraryExistenceChecker.java:
```diff
@@ -0,0 +1,66 @@
+     * @param otpDispatcherFunction a function that takes String as input and returns {@link OtpDispatcherResponse},
+     *                              example: {@link OtpDispatcher#sendOtpPlanRequest}.
+     */
+    public ItineraryExistenceChecker(Function<String, OtpDispatcherResponse> otpDispatcherFunction) {
+        if (otpDispatcherFunction == null) throw new NullPointerException();
+        this.otpDispatcherFunction = otpDispatcherFunction;
+    }
```
I'm not a huge fan of this constructor because it appears the sole purpose of it is to inject testing code. I think it'd be better to have the tests always use the `OtpDispatcher` and then properly mock some requests/responses using the `TestUtils.setupOtpMocks` method.

binh-dam-ibigroup replied:
Using mpckOtpServer as of https://github.com/ibi-group/otp-middleware/pull/77/commits/81d7fa65c0868e2811e6aa4889919b9e419ecb71

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryExistenceCheckerTest.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Test
+    public void testAllTripsExist() {
+        ItineraryExistenceChecker tripChecker = new ItineraryExistenceChecker(queryToResponse::get);
```
I can see how it does simplify testing the class, but it still seems a bit problematic. Is there a reason why this is preferred over using the `TestUtils.setupOtpMocks` method?

binh-dam-ibigroup replied:
I was trying to avoid setting up a mock server altogether, but sounds like you feel strongly about doing that instead.

binh-dam-ibigroup replied:
Using mockOtpServer as of https://github.com/ibi-group/otp-middleware/pull/77/commits/81d7fa65c0868e2811e6aa4889919b9e419ecb71

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java:
```diff
@@ -0,0 +1,252 @@
+    public static final long _2020_08_14__03_00_00_EDT = 1597388400000L; // Aug 14, 2020 3:00:00 AM
+
+    @Test
+    public void testGetQueriesFromDates() throws URISyntaxException {
```
Please add JavaDoc to each test case.

binh-dam-ibigroup replied:
Updated in https://github.com/ibi-group/otp-middleware/pull/77/commits/097c7f2bae3c9e29bcd6894de370cedefdaa3e12.

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java:
```diff
@@ -0,0 +1,252 @@
+        Assertions.assertEquals(itineraries.get(1), trip.itinerary);
+    }
+
+    private void testIsSameDay(String time, boolean isArrival, boolean expected, Long... startTimes) throws URISyntaxException {
```
This method along with all the methods that call it could be combined into a single ParameterizedTest. The test result looks a lot cleaner and if one of the test cases fails, the others will still be ran.

binh-dam-ibigroup replied:
Updated in https://github.com/ibi-group/otp-middleware/pull/77/commits/097c7f2bae3c9e29bcd6894de370cedefdaa3e12.

evansiroky reviewed (changes requested):
See various comments. 

Also, https://github.com/ibi-group/otp-middleware/pull/77#discussion_r491251522 is still something that has been unresolved. I am increasingly thinking that making OTP plan requests is the wrong way to go about checking for whether the itinerary is valid in the future. The OTP plan endpoint is not designed to have any comparison or checking abilities upon an itinerary that has been returned. Instead, it is portal to performing a shortest-path optimization given input parameters. 

I think a better solution would involve building in a new endpoint/set of calculations into OTP itself which would be able to calculate the itinerary validity. In doing so, OTP would have much greater detail since it would have all of the transit/road network information readily available. 

I'd hold off on further changes to this PR until we get on the same page with how to check for itinerary validity.

binh-dam-ibigroup committed:
refactor: Perform light refactorings for PR feedback.

binh-dam-ibigroup committed:
refactor(ItineraryExistenceCheckerTest): Use mock OTP server.

binh-dam-ibigroup committed:
refactor(ItineraryExistenceChecker): Merge class into ItineraryUtils.

binh-dam-ibigroup committed:
refactor(ItineraryUtilsTest): Use parametrized test for same-day test cases.

binh-dam-ibigroup committed:
refactor(ItineraryUtilsTest): Reuse static mock response.

binh-dam-ibigroup committed:
refactor(ItineraryUtilsTest): Add one more parametrized test.

br648 reviewed (approved):
Still looks good to me.

evansiroky commented on src/main/java/org/opentripplanner/middleware/otp/OtpDispatcherResponse.java:
```diff
@@ -90,4 +99,32 @@ public OtpDispatcherResponse clone() {
+            String requestTime = reqParams.get(TIME_PARAM);
+            if (requestDate != null && requestTime != null) {
+                // Get the zone id for this plan. Don't search for itineraries in this plan if the zone cannot be determined.
+                Optional<ZoneId> fromZoneId = getZoneIdForCoordinates(plan.from.lat, plan.from.lon);
```
I removed this ZoneId lookup in #75 as it turns out that this is not how OTP deals with timezones. Instead you'll want to use `DateTimeUtils. getOtpZoneId`.

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java:
```diff
@@ -0,0 +1,235 @@
+     */
+    public static List<String> getDatesToCheckItineraryExistence(MonitoredTrip trip, boolean checkAllDays) throws URISyntaxException {
+        List<String> result = new ArrayList<>();
+        ZoneId zoneId = trip.timezoneForTargetLocation();
```
I removed this ZoneId lookup in #75 as it turns out that this is not how OTP deals with timezones. Instead you'll want to use `DateTimeUtils. getOtpZoneId`.

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java:
```diff
@@ -0,0 +1,235 @@
+     * @param checkAllDays true if all days of the week are included regardless of the trip's monitored days.
+     * @return A list of date strings in YYYY-MM-DD format corresponding to each day of the week to monitor.
+     */
+    public static List<String> getDatesToCheckItineraryExistence(MonitoredTrip trip, boolean checkAllDays) throws URISyntaxException {
```
Line is too long, please break up.

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java:
```diff
@@ -0,0 +1,235 @@
+
+        // Start from the query date, if available.
+        // If there is no query date, start from today.
+        String queryDateString = params.getOrDefault(DATE_PARAM, DateTimeUtils.getStringFromDate(DateTimeUtils.nowAsLocalDate(), DEFAULT_DATE_FORMAT_PATTERN));
```
Line is too long, please break up.

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java:
```diff
@@ -0,0 +1,235 @@
+        LocalDate queryDate = DateTimeUtils.getDateFromString(queryDateString, DEFAULT_DATE_FORMAT_PATTERN);
+
+        // Get the trip time.
+        ZonedDateTime queryZonedDateTime = ZonedDateTime.of(queryDate, LocalTime.of(trip.tripTimeHour(), trip.tripTimeMinute()), zoneId);
```
Line is too long, please break up.

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java:
```diff
@@ -0,0 +1,377 @@
+    void testIsSameDay(SameDayTestCase testCase) throws URISyntaxException {
+        MonitoredTrip trip = makeTestTrip(testCase.isArrival);
+        // The time zone for trip (and testCase.tripTime) is US Eastern per trip location.
+        ZoneId zoneId = trip.timezoneForTargetLocation();
```
I removed this ZoneId lookup in #75 as it turns out that this is not how OTP deals with timezones. Instead you'll want to use `DateTimeUtils. getOtpZoneId`.

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java:
```diff
@@ -0,0 +1,377 @@
+            new SameDayTestCase(QUERY_TIME, _2020_08_13__03_00_00_EDT, false, true),
+            new SameDayTestCase(QUERY_TIME, _2020_08_13__23_59_59_EDT, false, true),
+            new SameDayTestCase(QUERY_TIME, _2020_08_14__02_59_59_EDT, false, true),
+            new SameDayTestCase("1:23", _2020_08_12__03_00_00_EDT, false, true),
+            new SameDayTestCase("1:23", _2020_08_12__23_59_59_EDT, false, true),
```
Since OTP shouldn't ever return a "depart at" itinerary that begins before the date + time request parameters, I don't believe such a case would occur. But if it did, I guess this is what I'd expect for being the same day though, so I guess it's fine to keep these test cases.

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java:
```diff
@@ -0,0 +1,377 @@
+            new SameDayTestCase("1:23", _2020_08_13__02_59_59_EDT, false, true),
+
+            // Not same-day departures
+            new SameDayTestCase(QUERY_TIME, _2020_08_12__23_59_59_EDT, false, false),
+            new SameDayTestCase(QUERY_TIME, _2020_08_13__02_59_59_EDT, false, false),
```
Since OTP shouldn't ever return a "depart at" itinerary that begins before the date + time request parameters, I don't believe such a case would occur. But if it did, I guess this is what I'd expect for being the same day though, so I guess it's fine to keep these test cases.

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java:
```diff
@@ -0,0 +1,377 @@
+            new SameDayTestCase(QUERY_TIME, _2020_08_12__23_59_59_EDT, false, false),
+            new SameDayTestCase(QUERY_TIME, _2020_08_13__02_59_59_EDT, false, false),
+            new SameDayTestCase(QUERY_TIME, _2020_08_14__03_00_00_EDT, false, false),
+            new SameDayTestCase("1:23", _2020_08_13__03_00_00_EDT, false, false),
```
This one threw me off a bit, but I now see that the line is drawn right at 3am, so this makes sense.

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java:
```diff
@@ -0,0 +1,377 @@
+
+            // Same-day arrivals
+            new SameDayTestCase(QUERY_TIME, _2020_08_13__03_00_00_EDT, true, true),
+            new SameDayTestCase(QUERY_TIME, _2020_08_13__23_59_59_EDT, true, true),
+            new SameDayTestCase(QUERY_TIME, _2020_08_14__02_59_59_EDT, true, true),
```
Since OTP shouldn't ever return a "arrive by" itinerary that ends after the date + time request parameters, I don't believe such a case would occur. But if it did, I guess this is what I'd expect for being the same day though, so I guess it's fine to keep these test cases.

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java:
```diff
@@ -0,0 +1,377 @@
+     * Data structure for the same-day test.
+     */
+    private static class SameDayTestCase {
+        public final boolean isArrival;
```
For consistency, please rename to `isArriveBy`

binh-dam-ibigroup replied:
Okay, this was the comment I missed. Updated in https://github.com/ibi-group/otp-middleware/pull/77/commits/36221dbf5c6ad2a52748e2995e4eb06c00be790a.

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java:
```diff
@@ -0,0 +1,377 @@
+    /**
+     * Helper method to create a bare-bones itinerary with start or end time.
+     */
+    private Itinerary simpleItinerary(Long startTime, boolean isArrival) {
```
Please rename `startTime` to `targetEpochMillis`.

binh-dam-ibigroup replied:
This was addressed in https://github.com/ibi-group/otp-middleware/pull/77/commits/78e31e7b50bcda047f65ddf1cbbbe506dbb8e455#diff-5a730e4259529c8d307574db916d01f209537dd2f60746a86a6e9b022ce8cca4R286.

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java:
```diff
@@ -0,0 +1,377 @@
+            this.isArrival = isArrival;
+            this.shouldBeSameDay = shouldBeSameDay;
+            this.timeOfDay = timeOfDay;
+            this.tripTime = tripTime;
```
Please always name a variable to have the specific time units the variable is supposed to represent. Consider naming this `tripTargetTimeEpochMillis`.

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java:
```diff
@@ -0,0 +1,377 @@
+            new SameDayTestCase("1:23", _2020_08_14__02_59_59_EDT, false, false),
+
+            // Same-day arrivals
+            new SameDayTestCase(QUERY_TIME, _2020_08_13__03_00_00_EDT, true, true),
```
This one is kind of tricky. Since it does end on the desired date, I suppose it makes sense. However, it implies that the trip has started on the previous day. If this ever comes up it's probably due to some really infrequent routes that run at odd times, so I guess it's fine to keep for now.

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java:
```diff
@@ -0,0 +1,377 @@
+            new SameDayTestCase(QUERY_TIME, _2020_08_14__02_59_59_EDT, true, true),
+            new SameDayTestCase("1:23", _2020_08_12__03_00_00_EDT, true, true),
+            new SameDayTestCase("1:23", _2020_08_12__23_59_59_EDT, true, true),
+            new SameDayTestCase("1:23", _2020_08_13__02_59_59_EDT, true, true),
```
Since OTP shouldn't ever return a "arrive by" itinerary that ends after the date + time request parameters, I don't believe such a case would occur. But if it did, I guess this is what I'd expect for being the same day though, so I guess it's fine to keep these test cases.

evansiroky commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java:
```diff
@@ -0,0 +1,377 @@
+            // Not same-day arrivals
+            new SameDayTestCase(QUERY_TIME, _2020_08_12__23_59_59_EDT, true, false),
+            new SameDayTestCase(QUERY_TIME, _2020_08_13__02_59_59_EDT, true, false),
+            new SameDayTestCase(QUERY_TIME, _2020_08_14__03_00_00_EDT, true, false),
```
Since OTP shouldn't ever return a "arrive by" itinerary that ends after the date + time request parameters, I don't believe such a case would occur. But if it did, I guess this is what I'd expect for being the same day though, so I guess it's fine to keep these test cases.

evansiroky replied:
This comment applies to the remaining 3 cases as well.

evansiroky reviewed (changes requested):
Looking good, just a few changes requested:

- some minor formatting items
- renaming some variables to note the type of time representation
- updating to the timezone finding methodology that was changed in the latest edits of #75

binh-dam-ibigroup committed:
Merge branch 'dev' into check-trip-existence

binh-dam-ibigroup committed:
refactor: Address PR comments.

binh-dam-ibigroup committed:
refactor(ItineraryUtilsTest): Replace magic millis with date/time objects.

evansiroky reviewed (approved):
Good to go after addressing https://github.com/ibi-group/otp-middleware/pull/77#discussion_r502725596.

binh-dam-ibigroup committed:
refactor(ItineraryUtilsTest): Rename one more variable.

landonreed commented on src/main/java/org/opentripplanner/middleware/controllers/api/ItineraryCheckController.java:
```diff
@@ -0,0 +1,120 @@
+ * This endpoint checks is used by web clients or other APIs to let users know
+ * whether a given itinerary exists on each day of the week.
+ */
+public class ItineraryCheckController implements Endpoint {
```
This design seems a little clunky to me. This entire controller does not need to exist. Here's what I propose:
1. In the UI, when a user clicks save trip, we send a request to createMonitoredTrip.
2. As part of the creation of that monitored trip, we kick off a checkItinerary job (Runnable), that gets tracked in a status object as part of the MonitoredTrip it sends back in the create trip response.
3. Meanwhile in the UI, we process the response from create trip and keep polling the get monitored trip endpoint to check the check itinerary status until it is complete. As itinerary existence is confirmed on the various days, that status gets updated.
4. Once the check is complete, the monitored trip status is fully updated and the user can proceed with confirming the trip/notification details.

landonreed replied:
Oh, actually the check already happens in `initializeTripAndSetVerifiedItinerary`. I need to dig into this a little more to understand why the separate controller exists then.

binh-dam-ibigroup replied:
This controller is for the UI to show which days are possible or not.

landonreed reviewed (changes requested):
I'm proposing a different way to to handle this itinerary check (see comment).

landonreed commented on src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java:
```diff
@@ -0,0 +1,243 @@
+        /** Whether all itineraries checked exist. */
+        public final boolean allItinerariesExist;
+        /**
+         * A map with the same keys as the input from checkAll,
```
What is `checkAll`? I need to read more, but I'm a bit suspect of this labeledQueries approach.

binh-dam-ibigroup replied:
`checkAll` determines whether all days of the week are checked. The UI needs to check all days to report which days are not available, whereas `monitoredtrip` PUT/POST only needs to checks trip existence on the days selected for monitoring.

landonreed committed:
refactor(itin-check): refactor main itinerary check methods/response

landonreed committed:
refactor(ItineraryExistence): add timestamp field; fix javadoc

binh-dam-ibigroup committed:
Merge pull request #89 from ibi-group/check-trip-existence-ltr

refactor(itin-check): refactor main itinerary check methods/response

binh-dam-ibigroup committed:
fix: Make tests pass again.

binh-dam-ibigroup committed:
refactor: Refactor tests.

binh-dam-ibigroup committed:
refactor: Reinstate checkAll flag to check itinerary on all days of week.

binh-dam-ibigroup committed:
refactor: Tweak comments.

binh-dam-ibigroup committed:
chore(swagger): Update version controlled swagger.

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java:
```diff
@@ -0,0 +1,166 @@
+    /**
+     * Overload used in normal conditions.
+     */
+    public static ItineraryExistence checkItineraryExistence(MonitoredTrip trip, boolean checkAllDays) throws URISyntaxException {
+        return checkItineraryExistence(trip, checkAllDays,false);
+    }
+
+    /**
+     * Variant of checkItineraryExistence used for tests,
+     * to ensure that the order of the dates to check matches the order of the mock OTP responses for those dates.
+     */
+    public static ItineraryExistence checkItineraryExistenceOrdered(MonitoredTrip trip, boolean checkAllDays) throws URISyntaxException {
+        return checkItineraryExistence(trip, checkAllDays,true);
+    }
```
This doesn't seem particularly helpful to make two new methods just to set a boolean argument. These could probably be removed.

binh-dam-ibigroup replied:
Probably the `ordered` variant doesn't need to be here. I'd rather keep the other one to hide the sorting argument, though.

evansiroky reviewed (approved):
Refactor looks good. There's just one item I don't feel strongly about.

binh-dam-ibigroup committed:
refactor(ItineraryUtils): Remove one method overload

binh-dam-ibigroup committed:
refactor: Address comments from #91 relevant to this PR.

landonreed committed:
refactor(itin-existence): refactor itinerary existence check

landonreed reviewed (changes requested):
See suggested changes in #93.

binh-dam-ibigroup committed:
Merge branch 'check-trip-existence' into check-trip-existence-ltr-2

binh-dam-ibigroup committed:
refactor: Make code compile and tests pass.

binh-dam-ibigroup committed:
Merge pull request #93 from ibi-group/check-trip-existence-ltr-2

Refactor itinerary existence check

binh-dam-ibigroup committed:
refactor(ItineraryExistence): Changes to make backend play well with client requests.

binh-dam-ibigroup committed:
refactor: Address #93 comments.

Remove full qualification, change response to 400 for trips not possible on certain days.

binh-dam-ibigroup committed:
chore(swagger): Update version controlled swagger.

binh-dam-ibigroup committed:
improvement(ItineraryExistence): Sort days of week in error message when trip not possible.

binh-dam-ibigroup committed:
docs(MonitoredTripController): Tweak endopint swagger doc.

binh-dam-ibigroup committed:
docs: Add JavaDoc, update swagger public tags.

landonreed commented on src/main/java/org/opentripplanner/middleware/models/ItineraryExistence.java:
```diff
@@ -0,0 +1,255 @@
+     */
+    @JsonIgnore
+    @BsonIgnore
+    public Collection<String> getInvalidDaysOfWeek() {
```
Since this is no longer returning formatted date strings, it might make more sense just to return a string message instead of a collection of Strings. In other words, the join the strings once again. I originally wrote it to return a collection of dates in case that was needed for some other purpose, but now it seems more specialized for just writing out a message.

binh-dam-ibigroup replied:
Updated in https://github.com/ibi-group/otp-middleware/pull/77/commits/d26baf619770dce9e11667961912b80261de090c..

landonreed replied:
Forgot the javadoc (!)

binh-dam-ibigroup replied:
Despair... Updated in https://github.com/ibi-group/otp-middleware/pull/77/commits/99a34ede08b01c3684b17a33315869c1e36ce3d4.

landonreed commented on src/main/java/org/opentripplanner/middleware/models/ItineraryExistence.java:
```diff
@@ -0,0 +1,255 @@
+         * Holds any matching itineraries (sorted by date) for the applicable day of the week.
+         * TODO: remove itineraries from list of returned fields in swagger.
+         */
+        public transient List<Itinerary> itineraries = new ArrayList<>();
```
Does this need to be public?

binh-dam-ibigroup replied:
This needs to be public for the tests in `ItineraryUtilsTest`.
Swagger will publish all fields in a type regardless of scope according to https://github.com/manusant/spark-swagger/blob/master/src/main/java/io/github/manusant/ss/factory/DefinitionsFactory.java#L36, so we can either leave it listed in the swagger doc for future use, or remove it during post-processing of the generated swagger. Which one should we do?

binh-dam-ibigroup replied:
I added code to remove these fields in the public docs in https://github.com/ibi-group/otp-middleware/pull/77/commits/d26baf619770dce9e11667961912b80261de090c.

landonreed commented on src/main/java/org/opentripplanner/middleware/models/ItineraryExistence.java:
```diff
@@ -0,0 +1,255 @@
+
+        /**
+         * Method and argument solely needed for accepting request contents that include a `valid` field
+         * (e.g. for checking a monitored trip with previous existence check record).
```
This doesn't make a ton of sense to me.

binh-dam-ibigroup replied:
Updated in https://github.com/ibi-group/otp-middleware/pull/77/commits/d26baf619770dce9e11667961912b80261de090c.

landonreed commented on src/main/resources/latest-spark-swagger-output.yaml:
```diff
@@ -859,6 +878,21 @@ definitions:
+        type: "array"
+        items:
+          type: "string"
+      itineraries:
```
Not a big deal for me, but the `isValid` JSON prop is not showing up here.

binh-dam-ibigroup replied:
Same reason as for the `itineraries` field. Although there is a getter and setter,`valid` is not a field, so spark-swagger just ignores it.

binh-dam-ibigroup replied:
I added the field in the public docs in https://github.com/ibi-group/otp-middleware/pull/77/commits/d26baf619770dce9e11667961912b80261de090c.

landonreed commented on src/main/resources/latest-spark-swagger-output.yaml:
```diff
@@ -964,6 +1005,36 @@ definitions:
+        items:
+          $ref: "#/definitions/OtpRequest"
+      referenceItinerary:
+        $ref: "#/definitions/Itinerary"
```
More non-blocking spark-swagger weirdness: doesn't seem like private fields should be shown.

binh-dam-ibigroup replied:
I added code to remove these fields in the public docs in https://github.com/ibi-group/otp-middleware/pull/77/commits/d26baf619770dce9e11667961912b80261de090c.

landonreed commented on src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java:
```diff
@@ -0,0 +1,241 @@
+     * Test case in which all itineraries exist and result.allCheckedDatesAreValid should be true.
+     */
+    @Test
+    public void testAllItinerariesExist() throws URISyntaxException {
```
Just a minor comment on test names: we don't usually call them `testMethodName`. I would prefer all these test names match our other syntax. E.g., `canCheckValidItineraries` or something.

binh-dam-ibigroup replied:
Updated in https://github.com/ibi-group/otp-middleware/pull/77/commits/a655afcdf5b8e76d68cfecf0e6ba3edb2a5d43fa.

landonreed reviewed (changes requested):
Looking great. Just a couple of final comments.

binh-dam-ibigroup committed:
refactor(ItineraryUtilsTest): Rename tests.

binh-dam-ibigroup committed:
refactor(ItinearyExistence): Postprocess fields in swagger, tweak check outcome msg and comments.

binh-dam-ibigroup committed:
docs(ItineraryExistence): Fix comment.

landonreed reviewed (approved):
Let's go ahead and merge. Future testing might suggest further changes, but I think is good for now!

binh-dam-ibigroup committed:
style(ItineraryExistence): Add one blank line
